### DISCUSSION TITLE
Convert getSharedSecret output to 32 bytes and ECDH_compute_key

### DISCUSSION THREAD

**Discussion by vxow** [paulmillr/noble-curves#114] [Q&A] [answered]
The shareKey generated by using p256.getSharedSecret does not match the sharekey generated by OpenSSL ECDH_compute_key, nor does it match the sharekey generated by using subarray(1).

**Comment by paulmillr** [comment]
I have no idea. noble multiplies point B by secret scalar A, which is the general definition of shared secret / ecdh. It must always be the same between all implementations.

Ensure the output is compressed / uncompressed depending on your needs.

https://github.com/paulmillr/noble-curves/blob/a1a7dc9cbf4ad0630339c085c81482cb2e0b685a/src/abstract/weierstrass.ts#L921-L926

**Reply by vxow** [reply]
Thanks for your answer, I need your help, c++ ECDH_compute_key returns 32 bytes of data, getSharedSecret returns 33 bytes after compression, and 65 bytes without compression, I tried subarray(1), but the result still does not match. There are no examples to exchange with C++.

**Reply by TobyEalden** [reply]
I'm seeing this too, but using `EVP_PKEY_derive_*` functions. C++ calculates a 32 byte key but `getSharedSecret` calculates 33 bytes.

Did you get to the bottom of it?

**Reply by paulmillr** [reply, marked as answer]
if it's 32 bytes, then it's either:

a. SLICED: `res.slice(1, 33)`
b. HASHED: `sha256(res.slice(1, 33))`
c. HKDFED: `hkdf_sha256(res.slice(1, 33))`
d. SLICED and HASHED/HKDFED: b or d, but replace `res.slice(1, 33)` with `decompRes.slice(1, 65)`, `res.slice(0, 33)` or `decompRes.slice(0, 65)`

There is no standard for ECDH. 32 bytes, 33 bytes, hashing or kdfing all make sense depending on ones needs. We provide the most flexible output, that can be converted to anything you'd like.

**Reply by TobyEalden** [reply]
`res.slice(1, 33)` did the trick - many thanks!

A case of a just enough knowledge to be dangerous I fear...
